### PR TITLE
AML AutoML Vision Classifier PowerSkill

### PR DESCRIPTION
**Pull request by shanepeckham**
## Purpose

* This Power Skill uses the AutoML trained vision model output from the [Azure Machine Learning Data Labelling project](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-create-labeling-projects)

## Does this introduce a breaking change?

```
[ ] Yes
[X ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
Full documented details on how to implement and test
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: Vision/AutoMLVisionClassifier/app.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+@app.on_event("startup")
+async def startup_event():
+    try:
+        if os.environ['GET_LATEST_MODEL'].lower() == "true":
+            logging.info(f"Download latest model")
+            if class_model.get_latest_model(os.environ['EXPERIMENT_NAME']):
+                class_model.load_classification_model('models/train_artifacts/')
+            else:
+                class_model.load_classification_model(os.environ['AZUREML_MODEL_DIR'])
```
Comment (bleroy) [inline Vision/AutoMLVisionClassifier/app.py:42]: Move environment variable names to constants next to the one you already have.

File: Vision/AutoMLVisionClassifier/powerskill/models.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        self.azureml_model_dir = azureml_model_dir
+
+    def get_workspace(self):
+        """
+        # Uses the config.json file to load the AML workspace
+        :return: An AML workspace instance
+        """
+
+        # Our AML config file
+        with open("/usr/src/api/config.json", "r") as json_file:
```
Comment (bleroy) [inline Vision/AutoMLVisionClassifier/powerskill/models.py:35]: Can we move that path to an environment variable with this value as default? Also, can we construct that path with a more x-plat-friendly method?

File: Vision/AutoMLVisionClassifier/powerskill/models.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                # Get the latest run
+                logging.info(f"Getting last best child run {tags['automl_best_child_run_id']}")
+                child_run = run.get(ws, tags['automl_best_child_run_id'])
+                metrics = run.get_metrics()
+                logging.info(f"Accuracy (class) {metrics['accuracy']}")
+                file_names = child_run.get_file_names()
+                if "train_artifacts/model.pt" in file_names:
+                    logging.info('Found a trained model.pt')
+                    child_run.download_files(prefix='train_artifacts',
+                                             output_directory='/usr/src/api/models')
```
Comment (bleroy) [inline Vision/AutoMLVisionClassifier/powerskill/models.py:79]: Same thing for this path.

**Review by bleroy — CHANGES_REQUESTED** [review]
Thanks for the contrib. I only have a few suggestions, looks great otherwise.

File: README.md [hunk 4]
```diff
@@ -28,6 +28,7 @@ This project provides the following custom skills:
 | [HocrGenerator](Vision/HocrGenerator/README.md) | transforms the result of OCR into the hOCR format. | Vision | ![C#](https://img.shields.io/badge/language-C%23-brightgreen) | ![functions](https://img.shields.io/badge/deploy-Functions-blue) | ARM Template |
 | [ImageClustering](Vision/ImageClusteringSkill/README.md) | Uses clustering to automatically group and label images | Vision | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
 | [ImageSimilarity](Vision/ImageSimilarity/README.md) | Uses ResNet to find the top-n most similar images | Vision | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
+| [AutoMLVisionClassifier](Vision/AutoMLVisionClassifier/README.md) | Gets your latest Data Labelling AML AutoML Vision model and runs inference on it | Vision | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply README.md:31]: Changed as suggested

File: Vision/AutoMLVisionClassifier/README.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+docker push yourcontainerregistry].azurecr.io/amlclassifier:v1
+```
+
+By default the Azure Web App will pull the pre-built image from:
+
+```bash
+shanepeckham/amlclassifier_powerskill:v1
+```
+
+*Note DockerHub rate limits so it is best to re-tag or rebuild this image and push it to your container registry*
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/README.md:71]: Added link and changed wording as suggested

File: Vision/AutoMLVisionClassifier/README.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+Run the following command to build the inference API container image:
+
+```bash
+docker build -t [container_registry_name.azurecr.io/vision_classifier:[your_tag] .  
+```
+
+Note, the prebuilt model only exists in the docker image `shanepeckham/amlclassifier_powerskill:v1` so if you need
+the prebuilt model, first pull this image and then retag it.
+
+The container will require the following variables set at runtime, namely:
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/README.md:82]: Removed

File: Vision/AutoMLVisionClassifier/README.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+Now we can test the container by running it locally with our variables:
+
+```bash
+docker run -it --rm -p 5000:5000 -e DEBUG=true -e KEY=YourSecretKeyCanBeAnything 
+-e AZUREML_MODEL_DIR=models  -e GET_LATEST_MODEL=True 
+-e EXPERIMENT_NAME=labeling_Training_xxxxxxxx 
+[container_registry_name.azurecr.io/vision_classifier:[your_tag]
+```
+
+Upon starting you will see a few messages, these can be ignored at startup. See below:
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/README.md:121]: Added more detail as suggested

File: Vision/AutoMLVisionClassifier/README.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+```bash
+Failure while loading azureml_run_type_providers. Failed to load entrypoint automl = azureml.train.automl.run:AutoMLRun._from_run_dto with exception cannot import name 'RunType' from 'azureml.automl.core._run.types' (/usr/local/lib/python3.7/site-packages/azureml/automl/core/_run/types.py).
+/usr/local/lib/python3.7/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
+```
+
+You should also see the following:
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/README.md:128]: Added as suggested

File: Vision/AutoMLVisionClassifier/README.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+### Skillset
+
+We will create two skillsets, one is the Custom Vision Cognitive Services service which will extract captions and
+run object detection, the other is our AML Classifier PowerSkill.
+
+Populate the following value and run the [Now we create the skill sets](notebooks/AML%20Classifier.ipynb#Now-we-create-the-skill-sets)
+
+```bash
+COGSVC_KEY = ""  #  This is your Cognitive Services key that resides in the same region as ACS (used to compare custom vision captions and object detection)
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/README.md:274]: It does not look like this is possible yet, only possible for data sources  -
 https://docs.microsoft.com/en-us/azure/search/search-howto-managed-identities-data-sources#:~:text=Managed%20identities%20is%20a%20feature%20that%20provides%20Azure,source%20through%20Azure%20role-based%20access%20control%20%28Azure%20RBAC%29.

File: Vision/AutoMLVisionClassifier/app.py (hunk 1 above)
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/app.py:42]: Implemented

File: Vision/AutoMLVisionClassifier/powerskill/models.py (hunk 2 above)
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/powerskill/models.py:35]: This path is specified by the dockerfile during build and will be constant - the platform is also constant.

File: Vision/AutoMLVisionClassifier/powerskill/models.py (hunk 3 above)
Comment (shanepeckham) [reply Vision/AutoMLVisionClassifier/powerskill/models.py:79]: Same comment as before, these paths are constants as is the underlying platform

**Comment by bleroy** [comment]
Thanks for the contribution!

**Pull request merged**